File: BarberApp_Nhom6_Be/app/models/review.py
from __future__ import annotations
from sqlalchemy import Integer, Text, DateTime, ForeignKey
from sqlalchemy.orm import Mapped, mapped_column, relationship
from datetime import datetime
from app.models.base import Base

class Review(Base):
    __tablename__ = "reviews"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    booking_id: Mapped[int] = mapped_column(ForeignKey("bookings.id", ondelete="CASCADE"))
    user_id: Mapped[int] = mapped_column(ForeignKey("users.id", ondelete="CASCADE"))
    rating: Mapped[int] = mapped_column(Integer)  # 1..5 (DDL đã có CHECK)
    comment: Mapped[str | None] = mapped_column(Text)
    created_at: Mapped[datetime | None] = mapped_column(DateTime, default=datetime.utcnow)

    booking: Mapped["Booking"] = relationship(back_populates="reviews")
    user: Mapped["User"] = relationship(back_populates="reviews")
    replies: Mapped[list["ReviewReply"]] = relationship(back_populates="review")


class ReviewReply(Base):
    __tablename__ = "review_replies"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    review_id: Mapped[int] = mapped_column(ForeignKey("reviews.id", ondelete="CASCADE"))
    admin_id: Mapped[int] = mapped_column(ForeignKey("users.id", ondelete="CASCADE"))
    reply: Mapped[str] = mapped_column(Text)
    created_at: Mapped[datetime | None] = mapped_column(DateTime, default=datetime.utcnow)

    review: Mapped["Review"] = relationship(back_populates="replies")

File: BarberApp_Nhom6_Fe/barberapp_nhom6_fe/lib/models/review.dart
int? _toIntOrNull(dynamic v) {
  if (v == null) return null;
  if (v is int) return v;
  if (v is num) return v.toInt();
  if (v is String) return int.tryParse(v);
  return null;
}
int _toInt(dynamic v) => _toIntOrNull(v) ?? 0;

class ReviewReply {
  final int id;
  final int reviewId;
  final int adminId;
  final String reply;
  final DateTime createdAt;

  ReviewReply({
    required this.id,
    required this.reviewId,
    required this.adminId,
    required this.reply,
    required this.createdAt,
  });

  factory ReviewReply.fromJson(Map<String, dynamic> j) => ReviewReply(
    id: _toInt(j['id']),
    reviewId: _toInt(j['review_id']),
    adminId: _toInt(j['admin_id']),
    reply: (j['reply'] ?? '').toString(),
    createdAt: DateTime.parse(j['created_at'] as String),
  );
}

class Review {
  final int id;
  final int bookingId;
  final int userId;
  final int rating;          // 1..5
  final String? comment;
  final DateTime createdAt;
  final List<ReviewReply> replies;

  Review({
    required this.id,
    required this.bookingId,
    required this.userId,
    required this.rating,
    this.comment,
    required this.createdAt,
    required this.replies,
  });

  factory Review.fromJson(Map<String, dynamic> j) => Review(
    id: _toInt(j['id']),
    bookingId: _toInt(j['booking_id']),
    userId: _toInt(j['user_id']),
    rating: _toInt(j['rating']),
    comment: j['comment'] as String?,
    createdAt: DateTime.parse(j['created_at'] as String),
    replies: (j['replies'] as List? ?? const [])
        .map((x) => ReviewReply.fromJson(Map<String, dynamic>.from(x as Map)))
        .toList(),
  );
}
